ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/HD/DiffSeq.hs

```diff
@@ -0,0 +1,289 @@
+                     !(Maybe SlotNo) -- ^ Right-most slot number
+  deriving stock (Generic, Show, Eq)
+  deriving anyclass (NoThunks)
+
+data Element (ts :: ToStoreKind) k v =
+    Element
+      {-# UNPACK #-} !SlotNo
+                     !(TableDiff ts k v)
+  deriving stock (Generic, Show, Eq)
+  deriving anyclass (NoThunks)
+
+
+-- | Length as a @'Sum'@.
+newtype Length = Length {unLength :: Int}
+  deriving stock (Generic, Show, Eq, Ord)
+  deriving newtype (Num)
+  deriving anyclass (NoThunks)
+  deriving Semigroup via Sum Int
+  deriving Monoid via Sum Int
+  deriving Group via Sum Int
+
+-- | Right-most slot number as a @'Max'@.
+newtype SlotNo = SlotNo {unSlotNo :: Slot.SlotNo}
+  deriving stock (Generic, Show, Eq, Ord)
+  deriving newtype (Num)
+  deriving anyclass (NoThunks)
+  deriving Semigroup via Max Slot.SlotNo
+  deriving Monoid via Max Slot.SlotNo
+
+{-------------------------------------------------------------------------------
+  Top-measuring
+-------------------------------------------------------------------------------}
+
+instance (Ord k, Eq v) => TopMeasured (TopMeasure ts k v) (Element ts k v) where
+  measureTop (Element _ d) = TopMeasure 1 d
+
+instance (Ord k, Eq v) => Semigroup (TopMeasure ts k v) where
+  TopMeasure len1 d1 <> TopMeasure len2 d2 =
+      TopMeasure (len1 <> len2) (d1 <> d2)
+
+instance (Ord k, Eq v) => Monoid (TopMeasure ts k v) where
+  mempty = TopMeasure mempty mempty
+
+instance (Ord k, Eq v) => Group (TopMeasure ts k v) where
+  invert (TopMeasure len d) =
+    TopMeasure (invert len) (invert d)
+
+{-------------------------------------------------------------------------------
+  Internal-measuring
+-------------------------------------------------------------------------------}
+
+instance Measured (InternalMeasure ts k v) (Element ts k v) where
+  measure (Element slotNo _d) = InternalMeasure 1 (Just slotNo)
+
+instance Semigroup (InternalMeasure ts k v) where
+  InternalMeasure len1 sl1 <> InternalMeasure len2 sl2 =
+    InternalMeasure (len1 <> len2) (sl1 <> sl2)
+
+instance Monoid (InternalMeasure ts k v) where
+  mempty = InternalMeasure mempty mempty
+
+{-------------------------------------------------------------------------------
+  Short-hands for type-class constraints
+-------------------------------------------------------------------------------}
+
+-- Short-hands for @'TopMeasured'@ and @'InternalMeasured'@ constraints for
+-- the @'DiffSeq'@-related datatypes.
+
+-- | Short-hand for @'Topmeasured'@.
+type TM (ts :: ToStoreKind) k v =
+  TopMeasured (TopMeasure ts k v) (Element ts k v)
+
+-- | Short-hand for @'InternalMeasured'@.
+type IM (ts :: ToStoreKind) k v =
+  InternalMeasured (InternalMeasure ts  k v) (Element ts k v)
+
+-- | Short-hand for @'SuperMeasured'@.
+type SM (ts :: ToStoreKind) k v =
+  (TM ts k v, IM ts k v)
+
+{-------------------------------------------------------------------------------
+  API: type classes
```

Comment:
These type classes seem nonsensical because they only have one instance (for the `DiffSeq` datatype above), but they could be an ingredient for generalising the `DiffSeq` datatype